src/btrfs.c

```diff
@@ -0,0 +1,183 @@
+/**
+ * collectd - src/btrfs.c
+ * Copyright (C) 2018       Martin Langlotz
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by the
+ * Free Software Foundation; only version 2 of the License is applicable.
+ *
+ * This program is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin St, Fifth Floor, Boston, MA  02110-1301 USA
+ *
+ * Authors:
+ *   Martin Langlotz < stackshadow at evilbrain . de >
+ **/
+
+#include "collectd.h"
+
+#include "common.h"
+#include "plugin.h"
+#include "utils_avltree.h"
+#include "utils_ignorelist.h"
+#include "utils_mount.h"
+
+#include <btrfs/ioctl.h>
+
+#define PLUGIN_NAME "btrfs"
+static const char *config_keys[] = {"Path"};
+static int config_keys_num = STATIC_ARRAY_SIZE(config_keys);
+static char *conf_btrfs_path;
+static c_avl_tree_t *conf_list_btrfs_paths = NULL;
+
+static int btrfs_avl_compare(const void *key, const void *value) { return 1; };
+
+static int btrfs_config(const char *key, const char *value) {
+
+  // create linked-list
+  if (conf_list_btrfs_paths == NULL) {
+    conf_list_btrfs_paths = c_avl_create(btrfs_avl_compare);
+  }
+
+  if (strcasecmp("Path", key) == 0) {
+
+    int ret = 0;
+
+    // get btrfs-path
+    conf_btrfs_path = strdup(value);
+    if (conf_btrfs_path == NULL) {
+      return 1;
+    }
+
+    ret = c_avl_insert(conf_list_btrfs_paths, conf_btrfs_path, conf_btrfs_path);
+    if (ret < 0) {
+      ERROR("[btrfs] ERROR: c_avl_insert\n");
+      exit(-1);
+    }
+
+    DEBUG("[btrfs] Set path: %s \n", conf_btrfs_path);
+  }
+
+  return 0;
+} /* int disk_config */
+
+static void btrfs_submit_value(const char *folder, const char *error,
+                               gauge_t value) {
+
+  // value
+  value_t tmp_value = {.gauge = value};
+  value_list_t vl = VALUE_LIST_INIT;
+
+  // create value-list
+  vl.values = &tmp_value;
+  vl.values_len = 1;
+
+  // create value
+  sstrncpy(vl.plugin, PLUGIN_NAME, sizeof(vl.plugin));
+  snprintf(vl.plugin_instance, sizeof(vl.plugin_instance), "%s#%s", folder,
```

Comment:
The `error` variable should go to `type_instance`. `plugin_instance` should be the same for same `folder`.